experimental/dds/tree2/src/feature-libraries/sequence-field/compose.ts

```diff
@@ -169,30 +172,99 @@ function composeMarks<TNodeChange>(
 ): Mark<TNodeChange> {
 	const nodeChange = composeChildChanges(baseMark.changes, newMark.changes, newRev, composeChild);
 
-	if (markIsTransient(newMark)) {
-		return withNodeChange(baseMark, nodeChange);
+	if (isTransientEffect(newMark)) {
+		const newDetachRevision = newMark.detach.revision ?? newMark.revision ?? newRev;
+		if (markEmptiesCells(baseMark)) {
+			if (isMoveDestination(newMark.attach) && isMoveSource(newMark.detach)) {
+				assert(isMoveSource(baseMark), "Unexpected mark type");
+				setEndpoint(
+					moveEffects,
+					CrossFieldTarget.Destination,
+					getEndpoint(baseMark, undefined),
+					baseMark.count,
+					{ revision: newDetachRevision, localId: newMark.detach.id },
+				);
+			}
+
+			// baseMark is a detach which cancels with the attach portion of the transient,
+			// so we are just left with the detach portion of the transient.
+			return withRevision(
+				withNodeChange({ ...newMark.detach, count: baseMark.count }, nodeChange),
+				newDetachRevision,
+			);
+		}
+
+		if (isTransientEffect(baseMark)) {
+			// TODO: Cancel marks even when there are node changes
+			if (
+				nodeChange === undefined &&
+				areEqualCellIds(getOutputCellId(newMark, newRev, revisionMetadata), baseMark.cellId)
+			) {
+				return { count: baseMark.count, cellId: baseMark.cellId };
+			}
+
+			// `newMark`'s attach portion cancels with `baseMark`'s detach portion.
+			const originalAttach = { ...baseMark.attach };
+			if (originalAttach.revision === undefined && baseMark.revision !== undefined) {
+				originalAttach.revision = baseMark.revision;
+			}
+
+			const finalDetach = { ...newMark.detach };
+			const detachRevision = finalDetach.revision ?? newMark.revision ?? newRev;
+			if (detachRevision !== undefined) {
+				finalDetach.revision = detachRevision;
+			}
+
+			return withNodeChange(
+				{
+					type: "Transient",
+					cellId: baseMark.cellId,
+					count: baseMark.count,
+					attach: originalAttach,
+					detach: finalDetach,
+				},
+				nodeChange,
+			);
+		}
+
+		return withRevision(withNodeChange(newMark, nodeChange), newRev);
 	}
-	if (markIsTransient(baseMark)) {
-		if (isInsert(newMark)) {
-			// TODO: Make `withNodeChange` preserve type information so we don't need to cast here
-			const nonTransient = withNodeChange(baseMark, nodeChange) as Insert<TNodeChange>;
-			delete nonTransient.transientDetach;
-			return nonTransient;
+	if (isTransientEffect(baseMark)) {
+		if (markFillsCells(newMark)) {
+			const baseAttachRevision = baseMark.attach.revision ?? baseMark.revision;
+			if (isMoveDestination(baseMark.attach) && isMoveSource(baseMark.detach)) {
+				assert(isMoveDestination(newMark), "Unexpected mark type");
+				setEndpoint(
+					moveEffects,
+					CrossFieldTarget.Source,
+					getEndpoint(newMark, newRev),
+					baseMark.count,
+					{ revision: baseAttachRevision, localId: baseMark.attach.id },
+				);
```

Comment:
The block is handling a scenario where the new change has "undone" some tail portion of a series of moves made in the base change. The start of the combined move chain is part of the base change, and the end of the move chain must point to that.

I do think there is a bug here, and we should be setting the endpoint to `getEndpoint(baseMark.attach)` instead of taking the ID directly from the base attach.

Reply:
Ah, right, that policy makes sense when there are multiple detaches for the same cell, but that's not the case here.

I agree `getEndpoint(baseMark.attach)` seems to make sense here.